src/marin/processing/classification/autoscaler.py

```diff
@@ -0,0 +1,456 @@
+                ping_received_ready, _ = ray.wait([ping_received], num_returns=1, timeout=1.0)
+
+                logger.debug("Ping wait result", extra={"actor": str(actor), "ready": bool(ping_received_ready)})
+                if ping_received_ready:
+                    try:
+                        result = ray.get(ping_received_ready, timeout=5.0)
+                        if result:
+                            logger.debug(f"Ping received from actor: {actor}")
+                            alive_actors.append(actor)
+                    except Exception as e:
+                        logger.error(f"Error received trying to ping actor: {e}")
+
+            self.actors = alive_actors
+            alive_actor_futures = {}
+            for actor, futures in self.actor_futures.items():
+                alive_actor_futures[actor] = futures
+
+            self.actor_futures = alive_actor_futures
+            available_actors = [
+                actor
+                for actor in self.actors
+                if self.max_tasks_per_actor is None or len(self.actor_futures.get(actor, [])) < self.max_tasks_per_actor
+            ]
+
+        if not available_actors:
+            return None
+
+        return min(available_actors, key=lambda a: len(self.actor_futures.get(a, [])))
+
+    def _dispatch_task(self, task: list[dict[str, Any]]) -> ray.ObjectRef:
+        """Dispatch a single task to an available actor."""
+
+        actor = self._get_least_loaded_actor()
+        if actor is None:
+            raise RuntimeError("No actors available")
+
+        with self.actor_task_metadata_lock:
+            current_load = len(self.actor_futures.get(actor, []))
+            if self.max_tasks_per_actor is not None and current_load >= self.max_tasks_per_actor:
+                raise RuntimeError("Actor saturated")
+
+            future = actor.__call__.remote(task)
+            self.actor_futures[actor] = [*self.actor_futures.get(actor, []), future]
+            self.future_to_actor[future] = actor
+            self.future_to_task[future] = task
+            logger.debug(f"Assigning to actor: {actor} for task: {task}")
+            return future
+
+    def _cleanup_completed_futures(self):
+        """Remove completed futures from tracking."""
+        with self.actor_task_metadata_lock:
+            for actor in self.actors:
+                if actor in self.actor_futures:
+                    # Filter out completed futures
+                    pending = []
+                    for future in self.actor_futures[actor]:
+                        ready, _ = ray.wait([future], timeout=0)
+                        if not ready:
+                            pending.append(future)
+                        else:
+                            self.future_to_actor.pop(future, None)
+                            self.future_to_task.pop(future, None)
+                    self.actor_futures[actor] = pending
+
+    def get_pool_stats(self) -> dict[str, Any]:
```

Comment:
Unused -- in general, I'd do a pass over here and remove any unused code. This is a little more complicated to read than it needs to be.